workspace/utils/rota.py

```diff
@@ -0,0 +1,98 @@
+from os import environ
+
+from apiclient import discovery
+from google.oauth2 import service_account
+
+
+class RotaReporter(abc.ABC):
+    def __init__(self, title: str, spreadsheet_id: str, sheet_range: str):
+        self.title = title
+        self.spreadsheet_id = spreadsheet_id
+        self.sheet_range = sheet_range
+        pass
+
+    def get_rota_data_from_sheet(self):  # pragma: no cover
+        credentials = service_account.Credentials.from_service_account_file(
+            environ["GCP_CREDENTIALS_PATH"],
+            scopes=["https://www.googleapis.com/auth/spreadsheets.readonly"],
+        )
+        service = discovery.build("sheets", "v4", credentials=credentials)
+        return (
+            service.spreadsheets()
+            .values()
+            .get(
+                spreadsheetId=self.spreadsheet_id,
+                range=self.sheet_range,
+            )
+            .execute()
+        )["values"]
+
+    def report(self):
+        rota = self.get_rota()
+        blocks = [self.get_header_block()]
+
+        today = date.today()
+        this_monday = today - timedelta(days=today.weekday())
+        blocks.append(
+            self.get_rota_block_for_week(rota, this_monday, this_or_next="this")
+        )
+
+        next_monday = this_monday + timedelta(days=7)
+        blocks.append(
+            self.get_rota_block_for_week(rota, next_monday, this_or_next="next")
+        )
+
+        blocks.append(self.get_block_linking_rota_spreadsheet())
+        return json.dumps(blocks, indent=2)
+
+    def get_rota(self):
+        rows = self.get_rota_data_from_sheet()
+        rota = self.convert_rota_data_to_dictionary(rows)
+        return rota
+
+    @abc.abstractmethod
+    def convert_rota_data_to_dictionary(self) -> dict:
+        """
+        Takes the rows returned from get_rota_data_from_sheet and converts them into a dictionary with dates as keys
+        """
+
+    def get_header_block(self):
+        return {
+            "type": "header",
+            "text": {
+                "type": "plain_text",
+                "text": self.title,
+            },
+        }
+
+    @abc.abstractmethod
+    def get_rota_text_for_week(self, rota: dict, monday: date, this_or_next: str):
+        """
+        Returns plain text reporting either the rota or a message saying no rota data was found
+        """
+
+    def get_rota_block_for_week(self, rota: dict, monday: date, this_or_next: str):
+        return {
+            "type": "section",
+            "text": {
+                "type": "mrkdwn",
+                "text": self.get_rota_text_for_week(rota, monday, this_or_next),
+            },
+        }
+
+    def get_block_linking_rota_spreadsheet(self):
+        return {
+            "type": "section",
+            "text": {
+                "type": "mrkdwn",
+                "text": f"<https://docs.google.com/spreadsheets/d/{self.spreadsheet_id}|Open rota spreadsheet>",
+            },
+        }
```

Comment:
You could simplify the blocks more with a function like

```
def get_text_block(text, block_type="section", text_type="mrkdwn"):
    return {
            "type": block_type,
            "text": {
                "type": text_type,
                "text": text,
            },
        }
```
and re-use it for the header, week and spreadsheet link blocks.

There are probably other places where a blocks helper utility could be used too (e.g. the project board reports) and the repo dashboards you're going to be looking at. (Not for this PR though)

Reply:
This makes a lot of sense - and gets rid of the need to pass a bunch of arguments through two functions! 
Adding ```utils/blocks.py``` to contain the ```get_text_block```; these might be renamed later when doing refactoring for the other jobs not in this PR.